File: backend/dineflow/DineFlow.Api/Services/AllergenDisclosure.cs
namespace DineFlow.Api.Services;

/// <summary>
/// One line of allergen disclosure and where it came from.
/// </summary>
/// <param name="Source">The dish, or the name of the modifier that carries it.</param>
/// <param name="Text">The wording the restaurant supplied, verbatim.</param>
public readonly record struct AllergenSource(string Source, string Text);

/// <summary>
/// Everything declared about a plate as it was actually ordered, dish and modifiers together.
/// </summary>
public sealed class AllergenDisclosure
{
    /// <summary>Named where each line came from, because "contains peanut" and "the satay sauce
    /// contains peanut" lead to different decisions — the second one can be removed.</summary>
    public IReadOnlyList<AllergenSource> Allergens { get; init; } = [];

    public IReadOnlyList<AllergenSource> MayContain { get; init; } = [];

    public IReadOnlyList<AllergenSource> CrossContact { get; init; } = [];

    /// <summary>True when nobody has declared anything at all, dish or modifier.</summary>
    public bool IsEmpty => Allergens.Count == 0 && MayContain.Count == 0 && CrossContact.Count == 0;
}

/// <summary>
/// Combines a dish's allergen declaration with those of the modifiers chosen for it.
///
/// <para>
/// A dish's declaration describes the dish as listed. Options had nowhere to record their own, so
/// adding satay sauce to a curry declared peanut-free left the panel the customer reads before
/// ordering saying exactly what it said before — the addition was invisible to the one check a
/// customer with an allergy actually performs.
/// </para>
///
/// <para>
/// The lines are kept separate and attributed rather than merged into one string. A customer who
/// reads "contains peanut (Satay sauce)" can deselect the sauce; one who reads "contains peanut"
/// can only put the dish down.
/// </para>
/// </summary>
public static class AllergenDisclosureBuilder
{
    /// <summary>How a dish names itself in an attribution.</summary>
    public const string DishSource = "This dish";

    /// <summary>
    /// The disclosure for a plate. Modifier lines follow the dish's, in the order chosen.
    /// </summary>
    /// <param name="modifiers">
    /// Name, allergens, may-contain and cross-contact for each selected option. A modifier that
    /// declares nothing contributes nothing — silence is not a claim of safety here, because the
    /// dish's own "not declared" already tells the customer the information is missing.
    /// </param>
    public static AllergenDisclosure Build(
        string? itemAllergens,
        string? itemMayContain,
        string? itemCrossContact,
        IEnumerable<(string Name, string? Allergens, string? MayContain, string? CrossContact)> modifiers)
    {
        var allergens = new List<AllergenSource>();
        var mayContain = new List<AllergenSource>();
        var crossContact = new List<AllergenSource>();

        Append(allergens, DishSource, itemAllergens);
        Append(mayContain, DishSource, itemMayContain);
        Append(crossContact, DishSource, itemCrossContact);

        foreach (var modifier in modifiers)
        {
            var name = string.IsNullOrWhiteSpace(modifier.Name) ? "A selected option" : modifier.Name.Trim();

            Append(allergens, name, modifier.Allergens);
            Append(mayContain, name, modifier.MayContain);
            Append(crossContact, name, modifier.CrossContact);
        }

        return new AllergenDisclosure
        {
            Allergens = allergens,
            MayContain = mayContain,
            CrossContact = crossContact
        };
    }

    private static void Append(List<AllergenSource> lines, string source, string? text)
    {
        if (!string.IsNullOrWhiteSpace(text))
        {
            lines.Add(new AllergenSource(source, text.Trim()));
        }
    }
}

File: frontend/dineflow-web/src/lib/allergenDisclosure.ts
/**
 * Everything declared about a plate as it is actually being ordered — the dish and the modifiers
 * chosen for it, together.
 *
 * <p>
 * A dish's declaration describes the dish as listed. Options had nowhere to record what they
 * themselves contain, so adding satay sauce to a curry declared peanut-free left the panel a
 * customer reads before ordering saying exactly what it said before. The one check an allergic
 * customer performs could not see the thing they had just added.
 * </p>
 *
 * <p>
 * Lines stay separate and attributed rather than merged into one sentence. Someone who reads
 * "contains peanut — Satay sauce" can deselect the sauce; someone who reads "contains peanut" can
 * only put the dish down.
 * </p>
 */

/** The dish itself, as opposed to one of its modifiers. */
export const dishSource = 'This dish'

export type AllergenLine = {
  /** `dishSource`, or the name of the modifier that carries this line. */
  source: string
  /** The restaurant's wording, verbatim. */
  text: string
}

export type PlateDisclosure = {
  allergens: AllergenLine[]
  mayContain: AllergenLine[]
  crossContact: AllergenLine[]
  /** True when a modifier contributed at least one line, so the panel can say the plate changed. */
  hasModifierDisclosure: boolean
}

type Declared = {
  allergens?: string | null
  mayContainAllergens?: string | null
  crossContactStatement?: string | null
}

type Modifier = Declared & { name: string }

function append(lines: AllergenLine[], source: string, text: string | null | undefined) {
  const trimmed = text?.trim()

  if (trimmed) {
    lines.push({ source, text: trimmed })
  }
}

export function buildPlateDisclosure(dish: Declared, modifiers: Modifier[]): PlateDisclosure {
  const allergens: AllergenLine[] = []
  const mayContain: AllergenLine[] = []
  const crossContact: AllergenLine[] = []

  append(allergens, dishSource, dish.allergens)
  append(mayContain, dishSource, dish.mayContainAllergens)
  append(crossContact, dishSource, dish.crossContactStatement)

  const dishLineCount = allergens.length + mayContain.length + crossContact.length

  for (const modifier of modifiers) {
    const source = modifier.name.trim() || 'A selected option'

    append(allergens, source, modifier.allergens)
    append(mayContain, source, modifier.mayContainAllergens)
    append(crossContact, source, modifier.crossContactStatement)
  }

  return {
    allergens,
    mayContain,
    crossContact,
    hasModifierDisclosure: allergens.length + mayContain.length + crossContact.length > dishLineCount,
  }
}

/**
 * One readable line per declaration. The dish's own lines are left unattributed — repeating "this
 * dish" on every row is noise — while a modifier's are named, because that is the part the customer
 * can still change.
 */
export function formatAllergenLines(lines: AllergenLine[]): string {
  return lines
    .map((line) => (line.source === dishSource ? line.text : `${line.text} (${line.source})`))
    .join('; ')
}

/**
 * The one line worth showing when the panel is folded away.
 *
 * <p>
 * The panel is collapsed by default because most of it is standing wording — how the kitchen is
 * laid out, what an order note cannot promise — that pushed the price and the options off a phone
 * screen. What must not be collapsed is the answer itself, so the summary carries it: an allergic
 * customer can decide from the closed panel and only opens it for the detail.
 * </p>
 *
 * <p>
 * Nothing declared is reported as such rather than as silence, which reads identically to a dish
 * confirmed to be free of everything.
 * </p>
 */
export function summariseDisclosure(disclosure: PlateDisclosure): {
  headline: string
  /** True when the restaurant has actually declared something, for styling and for tests. */
  declared: boolean
} {
  if (disclosure.allergens.length > 0) {
    return { headline: `Contains ${formatAllergenLines(disclosure.allergens)}`, declared: true }
  }

  if (disclosure.mayContain.length > 0) {
    return { headline: `May contain ${formatAllergenLines(disclosure.mayContain)}`, declared: true }
  }

  return { headline: 'Allergens not declared by the restaurant', declared: false }
}
